### PR TITLE
Adds logic to provide leniency for drag clicks

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

Ok so like, in byond. if you click on something and release the click without moving off its sprite, that counts as a "click"

If you however move your mouse off its sprite, or it does so on its own, any further releases count as dragdrops.

We already have logic for catching drag -> mouse back over sprite. The code I'm adding here basically is just a bit nicer about how exact you need to be

I'm adding tweakable values for how much your mouse is allowed to move (This isn't perfect currently, since if the target itself moves we don't care, need to work on this) in a short time period. If you meet these cases, we instead call it a click.

They're globs for now cause I don't have a good feel for em, will tweak over time. I want to reduce dumb clickspam and click dragging moving mobs/drifting items by accident.

Oh also fixed a bug in ai behaviors that came from not knowing what get_active_hand did

Fixes a monkey behavior of using their held item

It turns out it never worked, and the only reason it didn't runtime was because of an if() check that got passed the result of an invalid proc call. this fixes that

## Why It's Good For The Game

Spam clicking and drag clicking are stupid. they should be less stupid

## Changelog
:cl:
add: Adds a grace period for hit detection, between clicking ON someone, and then moving your mouse off before releasing the click. It's short, hopefully it helps with stupid
/:cl:

### CODE REVIEW

File: code/_onclick/drag_drop.dm [hunk 1]
```diff
@@ -10,14 +10,37 @@
...
 	over.MouseDrop_T(src,usr, params)
 	return
 
+// These two are globs for the purpose of testing, convert to defines when you're done
+// Distance in pixels that we consider "acceptable" from the initial click to the release
+// Note: this does not account for the position of the object, just where it was when we first started the timer
+GLOBAL_VAR_INIT(lieniency_distance, 10)
+// Accepted time in seconds between the initial click and drag release
+// Go higher then this and we just don't care anymore
+GLOBAL_VAR_INIT(lieniency_time, 0.1 SECONDS)
```
Comment (san7890) [inline code/_onclick/drag_drop.dm:27]:
```suggestion
GLOBAL_VAR_INIT(leniency_distance, 10)
// Accepted time in seconds between the initial click and drag release
// Go higher then this and we just don't care anymore
GLOBAL_VAR_INIT(leniency_time, 0.1 SECONDS)
```

File: code/_onclick/drag_drop.dm [hunk 2]
```diff
@@ -10,14 +10,37 @@
...
+/// Does the logic for checking if a drag counts as a click or not
+/// Returns true if it does, false otherwise
+/client/proc/check_drag_proximity(atom/dragging, atom/over, params)
+	if(dragging == over)
+		return TRUE
+	if(world.time - drag_start > GLOB.lieniency_time) // Time's up bestie
```
Comment (san7890) [inline code/_onclick/drag_drop.dm:33]:
```suggestion
	if(world.time - drag_start > GLOB.leniency_time) // Time's up bestie
```

File: code/_onclick/drag_drop.dm [hunk 3]
```diff
@@ -10,14 +10,37 @@
...
+		return FALSE
+	var/list/modifiers = params2list(params)
+	var/list/old_offsets = screen_loc_to_offset(LAZYACCESS(drag_details, SCREEN_LOC), view)
+	var/list/new_offsets = screen_loc_to_offset(LAZYACCESS(modifiers, SCREEN_LOC), view)
+
+	var/distance = sqrt(((old_offsets[1] - new_offsets[1]) ** 2) + ((old_offsets[2] - new_offsets[2]) ** 2))
+	if(distance > GLOB.lieniency_distance)
```
Comment (san7890) [inline code/_onclick/drag_drop.dm:40]:
```suggestion
	if(distance > GLOB.leniency_distance)
```

**Review by san7890 — COMMENTED** [review]
spellcheck

File: code/_onclick/drag_drop.dm (hunk 1 above)
Comment (LemonInTheDark) [reply code/_onclick/drag_drop.dm:27]: they're temporary so no

**Comment by LemonInTheDark** [comment]
~~I appreciate it but they're temporary vars, for testin~~

**Comment by Mothblocks** [comment]
update?

**Comment by LemonInTheDark** [comment]
Ok so like, it works, but it's kinda hard to test if it does what I want because it's not something people really notice in play.
I might uh, play a few rounds as a janitor or something and modify it between games until it feels right to me.  Not really sure how to approach this otherwise since the cases where it matters are sorta rare

**Comment by Mothblocks** [comment]
Log when it happens and see idk

**Comment by LemonInTheDark** [comment]
> Log when it happens and see idk

didn't even consider it, thanks bestie

**Comment by SkeletalElite** [comment]
I pretty frequently accidentally kick/punch someone when trying to fireman carry or strip when this is merged, is a little annoying.

**Comment by LemonInTheDark** [comment]
Should add a case for landing on a mob, since that's a lot of cases. I didn't think people dragged quickly enough for it to be a problem

**Comment by optimumtact** [comment]
Finish this please

**Comment by LemonInTheDark** [comment]
I think I'm happy with this, tho it can be tweaked in future. lemme undebug it

**Comment by LemonInTheDark** [comment]
Finished

**Pull request merged**